lib/rouge/lexers/php.rb

```diff
@@ -69,104 +100,273 @@ def self.analyze_text(text)
+        rule /#{NUMBER_HEX}/, Num::Hex
         rule /\d+/, Num::Integer
-        rule /'([^'\\]*(?:\\.[^'\\]*)*)'/, Str::Single
-        rule /`([^`\\]*(?:\\.[^`\\]*)*)`/, Str::Backtick
-        rule /"/, Str::Double, :string
+
+        rule /b?"/ do
+          token Str::Double
+          goto :double_quotes
+        end
+        rule /b?'/ do
+          token Str::Single
+          goto :single_quotes
+        end
+        rule /`/ do
+          token Str::Backtick
+          goto :backquote
+        end
+        rule /(b?<<<)(#{TABS_AND_SPACES})(["']?)(#{LABEL})(\3)(#{NEWLINE})/ do |m|
+          @label = m[4]
+          groups Operator, Text, Str::Heredoc, Name::Constant, Str::Heredoc, Text
+          goto "'" == m[3] ? :nowdoc : :heredoc
+        end
+
+        rule /#{NAMESPACED_LABEL}/ do |m|
+          token (
+            if m[0].include? '\\'
+              Name::Other
+            else
+              downcased = m[0].downcase
+              if self.class.keywords.include? downcased
+                Keyword
+              elsif self.builtins.include? downcased
+                Name::Builtin
+              else
+                Name::Other
+              end
+            end
+          )
+        end
+
+        rule /->/, Operator, :looking_for_property
+        # keep this last to not have precedence and create conflicts
+        rule /[~!%^&*+=\|:.<>\/?@-]+/, Operator
+
+        mixin :common
+        mixin :default
       end
 
-      state :classname do
-        rule /\s+/, Text
-        rule /[a-z_][\\\w]*/i, Name::Class, :pop!
+      # handles "short" (ie not surrounded by brackets)
+      # interpolation of array or string variable offset (eg: "$foo[...]")
+      state :var_offset do
+        rule /\[/, Punctuation
+        rule /]/, Punctuation, :pop!
+        rule /#{NAMESPACED_LABEL}/, Name::Constant
+        # non decimal integers are treated as strings
+        rule /#{NUMBER_BIN}|#{NUMBER_OCT}|#{NUMBER_HEX}/, Str
+        rule /0|[1-9]\d*/, Num::Integer
+
+        mixin :common
+        mixin :default
       end
 
-      state :funcname do
-        rule /[a-z_]\w*/i, Name::Function, :pop!
+      state :looking_for_property do
+        rule /->/, Operator
+        rule /#{LABEL}/, Name::Variable::Instance, :pop! # instance variable or method name (instance method call)
+        rule /#{WHITESPACE}/, Text
+        rule /./ do
+          pop!
+          restart!
+        end
       end
 
-      state :string do
-        rule /"/, Str::Double, :pop!
-        rule /[^\\{$"]+/, Str::Double
-        rule /\\([nrt\"$\\]|[0-7]{1,3}|x[0-9A-Fa-f]{1,2})/,
-          Str::Escape
-        rule /\$[a-zA-Z_][a-zA-Z0-9_]*(\[\S+\]|->[a-zA-Z_][a-zA-Z0-9_]*)?/, Name::Variable
+      state :looking_for_varname do
+        rule /#{LABEL}(?=[\[}])/ do
+          pop!
+          push :in_scripting
+          token Name::Variable
+        end
+        rule /./ do |m|
+          pop!
+          push :in_scripting
+          restart!
```

Comment:
Rather than insert new semantics to "undo" a match, you can use `rule(//) { ... }`